### ISSUE TITLE
Have an easier way to specify the uncertainties

### ISSUE THREAD

**Issue by eteq** [astropy/specutils#529]
This is partly  channeling an out-of-band discussion with @camipacifici . As it currently stands, passing uncertainties to `Spectrum1D` can be a bit awkward. It requires an `nddata` import which is often a non-intuitive step. So this issue is to discuss (and implement) a way to make this a bit easier on the user.

A few initial opinions from me to get the ball rolling:

* I'm strongly against "assume an array is stddev" or similar. If you ask 10 different astronomers, you'll probably get only 2 or 3 answers... But the different camps will swear to their graves that their way of representing uncertainties is the "right" one. So we want to keep the "you must say what kind of uncertainty it is" element of nddata.

So some possibilities:
1. In addition to supporting the current scheme, allow `Spectrum1D(spectral_axis=wavearr*u.angstrom, flux=fluxarr*funit, uncertainty=(uncarr*funit, 'stddev'))`. Then implement that as basically `uncertainty=getattr(astropy.nddata, t[1])(t[0])` at the top of `Spectrum1D.__init__`. Then the docs have to be very clear about how that mapping to `nddata` works.
2. add a `specutils.uncertainty` that imports nddata uncertainty objects (and gets imported into `specutils` at import time). Then at least the nddata import is not necessary.

I favor 1 over 2 here, but both seem straightforward so I thought I'd mention for discussion...

cc @crawfordsm @keflavich

**Comment by keflavich** [comment]
I actually like (2) a lot.  I don't like the idea of specifying complex object definitions as the tuples of their arguments, since it becomes difficult to track what's going on.  I like the idea of having the standard specutils import be something like (in pseudocorrect code...):
```
from specutils import Spectrum1D, StandardDeviationUncertainty
```
or
```
from specutils import Spectrum1D, uncertainty as spectral_uncertainty
```
or something on those lines.

I am opposed to this sort of construct:
`uncertainty=(uncarr*funit, 'stddev'))`
because even though the name is physically explicit, the construct being built under the hood is hidden and implicit.

**Comment by nmearl** [comment]
I have to agree with @keflavich: I really don't like the idea of having to construct a tuple. There is subtle disconnect from a user perspective about having to compose a loose object with no formal definition (i.e. there's nothing about a tuple that specifies the first argument must be a quantity, and the second around must be a string containing the reference to the type of uncertainty).

I'd say either (2), or break the keyword argument out in the `Spectrum1D` case and require that a `uncert_type` be passed in if `uncertainty` is defined.

```python
spec = Spectrum1D(flux=my_flux, spectral_axis=my_wavelength, uncertainty=my_uncerts, uncert_type='stddev')
```

**Comment by camipacifici** [comment]
The problem I have with the uncertainty is finding enough information to create the correct uncertainty object. Basically, if I start from the specutils documentation (https://specutils.readthedocs.io/en/stable/spectrum1d.html), there is a section "Including uncertainties", but at that page the documentation only says that "Uncertainties are a special subclass of NDData". So I click on the NDData link, but I do not find instructions of how to define the uncertainty object. Only after a few other searches and clicks, I finally get to this page http://docs.astropy.org/en/stable/nddata/ where the section "Image arithmetic, including uncertainty" tells me the options I have to define the Spectrum1D.uncertainty object. I think that including in the first page (https://specutils.readthedocs.io/en/stable/spectrum1d.html) the details that are specified in the right NDData page (http://docs.astropy.org/en/stable/nddata/) can be a solution. Simply adding this sentence "Three uncertainty types are supported: variance (VarianceUncertainty), standard deviation (StdDevUncertainty) and inverse variance (InverseVariance)." could clarify how to create the uncertainty.

@nmearl solution is very intuitive! However, it still requires the documentation to be clearer on the uncertainty options the user has.

**Comment by aburgasser** [comment]
The current framework for uncertainties is pretty awkward; for example, if I read in a spectrum from a file and want to plot the uncertainties and flux, it isn't immediately intuitive:

from astropy.io import fits
from astropy import units as u
from astropy.nddata import StdDevUncertainty
from matplotlib import pyplot as plt
from specutils import Spectrum1D

f = fits.open('some fits file with wave,flux,unc axes')
default_wave_unit = u.micron
default_flux_unit = u.erg/u.cm/u.cm/u.s/u.micron
wave,flux,unc = f[0].data
f.close()
spec = Spectrum1D(spectral_axis=wave*default_wave_unit, flux=flux*default_flux_unit, uncertainty=StdDevUncertainty(unc)) 
plt.plot(spec.spectral_axis,spec.flux)
plt.plot(spec.spectral_axis,spec.uncertainty)

the last line crashes, as it inteprets uncertainty as a single dimension "thing"; but this works:

plt.plot(spec.spectral_axis,spec.uncertainty.array)

And since spec.uncertainty.array also doesn't have units, the correct statement would really be:

plt.plot(spec.spectral_axis,spec.uncertainty.array*spec.flux.unit)

which is even more awkward.

I would suggest either making sure spec.uncertainty is interpreted as an array or generate a new variable in the class that explicitly links to the uncertainty.array with the right units so that it is treated on par with the other dimensions

**Comment by dhomeier** [comment]
> the last line crashes, as it inteprets uncertainty as a single dimension "thing"; but this works:
> 
> `plt.plot(spec.spectral_axis,spec.uncertainty.array) `
> And since spec.uncertainty.array also doesn't have units, the correct statement would really be:
> 
> `plt.plot(spec.spectral_axis,spec.uncertainty.array*spec.flux.unit) `
> which is even more awkward.
> 
> I would suggest either making sure spec.uncertainty is interpreted as an array or generate a new variable in the class that explicitly links to the uncertainty.array with the right units so that it is treated on par with the other dimensions

Uncertainty is an `astropy.nddata` class, so modifying it is a bit outside of this scope - we couldsubclass the `NDUncertainty` classes, but making it directly return a `Quantity` would likely break some things. For the time being, there is a slightly less ugly alternative:
```
plt.plot(spec.spectral_axis, spec.uncertainty.quantity)
```

**Comment by dhomeier** [comment]
Having also just pondered more convenient options to specify the uncertainty, my idea was simply an additional `kwarg` like `uncertainty_type='std'`, which is a bit lengthy to type, but avoids the awkward tuple construction and additional imports.

**Comment by aburgasser** [comment]
perhaps a more "internal" solution would be to assign this to another class variable, say "errors" = uncertainty.quantity ...?

**Comment by aburgasser** [comment]
maybe "flux_uncertainty" would be better

**Comment by nmearl** [comment]
> Uncertainty is an astropy.nddata class, so modifying it is a bit outside of this scope - we could subclass the NDUncertainty classes, but making it directly return a Quantity would likely break some things.

Yes, we certainly do not want to be maintaining custom subclasses for package-specific purposes. However, we have toyed with the idea of making `NDUncertainty` subclasses inherit e.g. `Quantity`, which would mean they could be handled as arrays without the extra attribute accessor.

> Having also just pondered more convenient options to specify the uncertainty, my idea was simply an additional kwarg like uncertainty_type='std', which is a bit lengthy to type, but avoids the awkward tuple construction and additional imports.

Yes, this was suggested a [couple comments up](https://github.com/astropy/specutils/issues/529#issuecomment-538062881), and seems the most probably and straight forward way of avoiding having the user learn the `Uncertainty` machinery unless they really want to.

> perhaps a more "internal" solution would be to assign this to another class variable, say "errors" = uncertainty.quantity ...?

Not a bad idea for an easy shortcut for users not really interested in learning the nuances of the uncertainty handling.

**Comment by dhomeier** [comment]
> > perhaps a more "internal" solution would be to assign this to another class variable, say "errors" = uncertainty.quantity ...?
> 
> Not a bad idea for an easy shortcut for users not really interested in learning the nuances of the uncertainty handling.

Maybe relevant: https://github.com/astropy/astropy/issues/10128
